aptos-move/aptos-vm/src/data_cache.rs

```diff
@@ -235,3 +233,30 @@ impl<S: StateView> AsMoveResolver<S> for S {
         StorageAdapter::new(self)
     }
 }
+
+impl<'a, S: StateView> StateValueMetadataResolver for StorageAdapter<'a, S> {
+    fn get_state_value_metadata(
+        &self,
+        state_key: &StateKey,
+    ) -> anyhow::Result<Option<Option<StateValueMetadata>>> {
+        let maybe_state_value = self.state_store.get_state_value(state_key)?;
+        Ok(maybe_state_value.map(|state_value| state_value.into_metadata()))
```

Comment:
So the outer Option is because of the maybe_state_value being empty, and inside is if the into_metadata doesn't succeed? can't we map_err to storage error on the outer one here to convert it to the result? (seems to be the way the caller uses it)?

Reply:
Not really: the outer thing can be None if the state value does not exist: in this case for storage refunds this is acceptable. If inner does not exist, then there is no metadata associated, so we kind of need both...

I guess using named enums instead of option would be nicer in general😄